### PR TITLE
[MRG] Add subclass of OptimumTrajectory that slows down movement near obstacles

### PR DESCRIPTION
**Pull request by maotto**
preliminary puil-request. As discussed, I will add a subclass instead. Also I did minor changes that aren't tested. Check back next week.

### CODE REVIEW

**Comment by maotto** [comment]
subclass added as announced. Also fixed an issue with the plotting function in the base class

File: bolero/environment/optimum_trajectory.py [hunk 1]
```diff
@@ -294,5 +294,116 @@ def plot(self, ax):
         if self.obstacles is not None:
             from matplotlib.patches import Circle
             for obstacle in self.obstacles:
-                ax.add_patch(Circle(obstacle, self.obstacle_dist, ec="none",
+                ax.add_patch(Circle(np.asarray(obstacle).copy(), self.obstacle_dist, ec="none",
```
Comment (AlexanderFabisch) [inline bolero/environment/optimum_trajectory.py]: Why is that required?

File: bolero/environment/optimum_trajectory.py [hunk 2]
```diff
@@ -294,5 +294,116 @@ def plot(self, ax):
...
                                     color="r"))
+
+
+class OptimumTrajectoryCurbingObstacles(OptimumTrajectory):
+    """Optimize a trajectory according to some criteria. Passing by obstacles slows down or stops the movement
```
Comment (AlexanderFabisch) [inline bolero/environment/optimum_trajectory.py]: The first sentence should describe the whole class briefly.

File: bolero/environment/optimum_trajectory.py [hunk 3]
```diff
@@ -294,5 +294,116 @@ def plot(self, ax):
...
+        super(OptimumTrajectoryCurbingObstacles, self).init()
+        self.X *= np.nan
+        self.Xd *= np.nan
+        self.Xdd *= np.nan
+
+    def reset(self):
+        """Reset state of the environment."""
+        super(OptimumTrajectoryCurbingObstacles, self).reset()
+        self.damping = 0
+        self.X *= np.nan
```
Comment (AlexanderFabisch) [inline bolero/environment/optimum_trajectory.py]: Why is it only here and not in the super class?

File: bolero/environment/optimum_trajectory.py [hunk 4]
```diff
@@ -294,5 +294,116 @@ def plot(self, ax):
...
+                 penalty_start_dist=0.0, penalty_goal_dist=0.0,
+                 penalty_vel=0.0, penalty_acc=0.0, penalty_obstacle=0.0,
+                 log_to_file=False, log_to_stdout=False):
+        super(OptimumTrajectoryCurbingObstacles, self).__init__(
+            x0=x0, g=g, execution_time=execution_time, dt=dt, obstacles=obstacles,
+            obstacle_dist=obstacle_dist, penalty_start_dist=penalty_start_dist,
+            penalty_goal_dist=penalty_goal_dist, penalty_vel=penalty_vel, penalty_acc=penalty_acc,
+            penalty_obstacle=penalty_obstacle, log_to_file=log_to_file, log_to_stdout=log_to_stdout)
+        self.curbing_obstacles = curbing_obstacles
+        self.damping = 0
```
Comment (AlexanderFabisch) [inline bolero/environment/optimum_trajectory.py]: would be better to set this in `init()`, because it is not set by the user of this class

File: bolero/environment/optimum_trajectory.py [hunk 5]
```diff
@@ -294,5 +294,116 @@ def plot(self, ax):
...
+
+    def init(self):
+        """Initialize environment."""
+        super(OptimumTrajectoryCurbingObstacles, self).init()
+        self.X *= np.nan
```
Comment (AlexanderFabisch) [inline bolero/environment/optimum_trajectory.py]: Why do you have to do it in `init()` and in `reset()`? `reset()` should be executed before each episode, also before the first episode.

File: bolero/environment/optimum_trajectory.py [hunk 6]
```diff
@@ -294,5 +294,116 @@ def plot(self, ax):
...
+        ----------
+        values : array,
+            Inputs for the environment: positions, velocities and accelerations
+            in that order, e.g. for n_task_dims=2 the order would be xxvvaa
+        """
+        if self.damping:
+            # can only occur after step action, thus self.t > 0
+            if self.t == 0:
+                raise ValueError
+            new_value_weight = max(0, (1-self.damping))
```
Comment (AlexanderFabisch) [inline bolero/environment/optimum_trajectory.py]: PEP8

File: bolero/environment/optimum_trajectory.py [hunk 7]
```diff
@@ -294,5 +294,116 @@ def plot(self, ax):
...
+    def set_inputs(self, values):
+        """Set environment inputs, e.g. next action.
+
+        Parameters
+        ----------
+        values : array,
+            Inputs for the environment: positions, velocities and accelerations
+            in that order, e.g. for n_task_dims=2 the order would be xxvvaa
+        """
+        if self.damping:
```
Comment (AlexanderFabisch) [inline bolero/environment/optimum_trajectory.py]: This looks like something that should be done in `step_action()`. `set_inputs()` should usually only store the input values and do nothing else.

File: bolero/environment/optimum_trajectory.py (hunk 2 above)
Comment (AlexanderFabisch) [inline bolero/environment/optimum_trajectory.py]: What does "passing obstacles" mean by the way? Does it mean you are in the obstacle or close to the obstacle?

**Comment by AlexanderFabisch** [comment]
The class should also be added here:
* `__init__.py` of `environment`: https://github.com/rock-learning/bolero/blob/master/bolero/environment/__init__.py
* Python API documentation: https://github.com/rock-learning/bolero/blob/master/doc/source/api_py.rst

**Comment by AlexanderFabisch** [comment]
@maotto Are you still working on this?

File: bolero/environment/optimum_trajectory.py (hunk 1 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: This is necessary, when obstacle positions are changed but the user wants to control when the plot is updated.

I'll request merging this class later, which allows adapting the task accordingly:
class ContextualParetoOptimalTrajectory(OptimumTrajectory, ContextualEnvironment):
    def get_num_context_dims(self):
        return self.obstacles.size

    def get_context(self):
        return self.obstacles.ravel()

    def request_context(self, obstacles):
        if obstacles is None:
            return self.get_context()
        self.obstacles = np.reshape(obstacles, (-1, len(self.x0)))

File: bolero/environment/optimum_trajectory.py (hunk 2 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: You mean it should only be one sentence in the first line? Here's an adapted version:
**"Optimize a trajectory according to some criteria while movements are slowed down or stopped in the vicinity of obstacles"**

File: bolero/environment/optimum_trajectory.py (hunk 3 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: I will move it

File: bolero/environment/optimum_trajectory.py (hunk 4 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: To avoid an instance attribute definition outside the __init__ method. I can move it

File: bolero/environment/optimum_trajectory.py (hunk 5 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: 1. If reset is not called although it should be, old data would remain.
2. numpy.empty does not initialize the entries, so the wrong usage can easily go undetected

If we want to avoid redundant code, I suggest asserting that the preallocated values to be overwritten are NaNs.
What do you prefer?

File: bolero/environment/optimum_trajectory.py (hunk 7 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: Another instance attribute would be necessary, i.e., more memory, more code and a higher risk that something goes wrong. But I agree that step action is the right place to change the state represented by X, Xd and Xdd, so I'll change it.

File: bolero/environment/optimum_trajectory.py (hunk 2 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: The effect begins when a target position is inside the obstacle_dist radius. Our current documentation of this parameter suggests that obstacles are just points (one cannot be inside the obstacle) and this is the distance to be kept. To increase clarity, I propose replacing "Passing by" with being **"in the vicinity of obstacles"** (see above).
However, to activate the feature, the user needs to change the parameter *curbing_obstacles*, which has a documentation that explains this.

**Comment by maotto** [comment]
> @maotto Are you still working on this?

Yes, but it has low priority (like ~ one update per month)

File: bolero/environment/optimum_trajectory.py (hunk 4 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: What if I assign `self.damping = 0` *only* in the reset method? In combination with an informative exception, a forgotten call to reset would be detected at the first run..

File: bolero/environment/optimum_trajectory.py (hunk 7 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: actually, set_inputs of the base class behaves the same and adapts the state already. For consistency, I would then rather not change it here.

File: bolero/environment/optimum_trajectory.py (hunk 5 above)
Comment (AlexanderFabisch) [reply bolero/environment/optimum_trajectory.py]: > If we want to avoid redundant code, I suggest asserting that the preallocated values to be overwritten are NaNs.

Yes, I think that is the correct way to detect these errors.

File: bolero/environment/optimum_trajectory.py (hunk 2 above)
Comment (AlexanderFabisch) [reply bolero/environment/optimum_trajectory.py]: > You mean it should only be one sentence in the first line?

Yes, that is what I meant. That suggestion is too long though. :)

What about "Search trajectories with several criteria and curbing near obstacles.", followed by a more detailed description in the next paragraph?

File: bolero/environment/optimum_trajectory.py [hunk 8]
```diff
@@ -294,5 +294,116 @@ def plot(self, ax):
...
+            # can only occur after step action, thus self.t > 0
+            if self.t == 0:
+                raise ValueError
```
Comment (AlexanderFabisch) [inline bolero/environment/optimum_trajectory.py]: You would raise a RuntimError here. Also please use syntax that is compatible to Python 3: `raise RuntimeError("bla bla bla")`

File: bolero/environment/optimum_trajectory.py (hunk 7 above)
Comment (AlexanderFabisch) [reply bolero/environment/optimum_trajectory.py]: > Another instance attribute would be necessary, i.e., more memory

That would just be an array of size `n_task_dims * 3`, right? I think that is neglectable.

> actually, set_inputs of the base class behaves the same and adapts the state already. For consistency, I would then rather not change it here.

Yes, that is not very good, either. But that one is already in master unfortunately. :) This implementation is even more complicated and it is definitively not what you expect from a setter or getter function.

File: bolero/environment/optimum_trajectory.py (hunk 1 above)
Comment (AlexanderFabisch) [reply bolero/environment/optimum_trajectory.py]: OK, but could you replace it by `np.copy(obstacle)`? I think `np.asarray(obstacle).copy()` copies obstacle twice if obstacle is not a numpy array.

File: bolero/environment/optimum_trajectory.py (hunk 2 above)
Comment (AlexanderFabisch) [reply bolero/environment/optimum_trajectory.py]: > The effect begins when a target position is inside the obstacle_dist radius. Our current documentation of this parameter suggests that obstacles are just points (one cannot be inside the obstacle) and this is the distance to be kept.

I always thought of it as circular obstacles.

> However, to activate the feature, the user needs to change the parameter _curbing_obstacles_, which has a documentation that explains this.

Then it should be clearly documented before the list of parameters in the docstring that you have to set this parameter. Otherwise an unexperienced user will wonder why the environment does not behave differently. An alternative would be to set a reasonable default value != 0.

Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: > I always thought of it as circular obstacles.

OK. We could change "Distance that should be kept to the obstacles" to "Radius of the obstacles that should be avoided" to reflect this. (see docstring for obstacle_dist). Then I would also change my docs from "Passing by" to "Passing through"

> a reasonable default value != 0.
Yes, the zero default was a remainder from when I implemented just an adaptation of the base class. I think 0.5 is reasonable: a single obstacle slows down, and passing trough al least two obstacles at the same time (when they overlap) will stop regardless of the inputs.

File: bolero/environment/optimum_trajectory.py (hunk 1 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: OK

File: bolero/environment/optimum_trajectory.py (hunk 2 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: Good. Reflecting our decision on the [non-point nature](https://github.com/rock-learning/bolero/pull/69#discussion_r219422851) of obstacles, I propose
"Search trajectories with several criteria and curbing *when passing through* obstacles."

File: bolero/environment/optimum_trajectory.py (hunk 7 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: alright

File: bolero/environment/optimum_trajectory.py (hunk 8 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: Is it just "bla bla" is python 2? ;-)
OK, I'll fix it.

File: bolero/environment/optimum_trajectory.py (hunk 2 above)
Comment (maotto) [reply bolero/environment/optimum_trajectory.py]: the pep8 79 char limit made me rethink: 
"Search trajectories with several criteria and curbing within obstacles."

**Comment by AlexanderFabisch** [comment]
> from my side, it's ready to be merged.
> I moved the state changes in the base class from set_inputs to step_action, which actually reduced code redundancy for the non-damping case and it makes more sense :)

Sorry for my nitpicking and the delay, but I think the code looks really good now. Just one more question: could you add at least one reasonable unit test for this class?

**Comment by AlexanderFabisch** [comment]
Looks good. Can you rebase on master?

**Comment by AlexanderFabisch** [comment]
TODO list for me:

- [ ] fix Travis CI
- [ ] update NEWS.md

**Comment by AlexanderFabisch** [comment]
see #88
